build is broken with protobuf

I just bumped rules_go to 59327146e5395cd1773bab0107d974f660cc0852 in order to get at Go 1.9.4 and my build has started breaking with:

```
(07:15:53) INFO: Found 95 targets and 12 test targets...
(07:16:01) ERROR: /home/travis/bazel/outbase/external/com_github_golang_protobuf/proto/BUILD.bazel:3:1: GoCompile external/com_github_golang_protobuf/proto/linux_amd64_race_stripped/go_default_library~/github.com/golang/protobuf/proto.a failed (Exit 1): compile failed: error executing command 
  (cd /home/travis/bazel/outbase/execroot/founding && \
  exec env - \
  bazel-out/host/bin/external/io_bazel_rules_go/go/tools/builders/linux_amd64_stripped/compile -go external/go_sdk/bin/go -root_file 'bazel-out/host/bin/external/io_bazel_rules_go/linux_amd64_stripped/stdlib~/ROOT' -goos linux -goarch amd64 '-cgo=1' -compiler_path /usr/bin -cc /usr/bin/gcc -cpp_flag -U_FORTIFY_SOURCE -cpp_flag -fstack-protector -cpp_flag -B/usr/bin -cpp_flag -B/usr/bin -cpp_flag -Wunused-but-set-parameter -cpp_flag -Wno-free-nonheap-object -cpp_flag -fno-omit-frame-pointer -cpp_flag -g0 -cpp_flag -O2 -cpp_flag '-D_FORTIFY_SOURCE=1' -cpp_flag -DNDEBUG -cpp_flag -ffunction-sections -cpp_flag -fdata-sections -cpp_flag -g0 -ld_flag -lstdc++ -ld_flag -lm -ld_flag '-fuse-ld=gold' -ld_flag -Wl,-no-as-needed -ld_flag -Wl,-z,relro,-z,now -ld_flag -B/usr/bin -ld_flag -B/usr/bin -ld_flag -pass-exit-codes -package_list external/go_sdk/packages.txt -src external/com_github_golang_protobuf/proto/clone.go -src external/com_github_golang_protobuf/proto/decode.go -src external/com_github_golang_protobuf/proto/encode.go -src external/com_github_golang_protobuf/proto/equal.go -src external/com_github_golang_protobuf/proto/extensions.go -src external/com_github_golang_protobuf/proto/lib.go -src external/com_github_golang_protobuf/proto/message_set.go -src external/com_github_golang_protobuf/proto/pointer_unsafe.go -src external/com_github_golang_protobuf/proto/properties.go -src external/com_github_golang_protobuf/proto/text.go -src external/com_github_golang_protobuf/proto/text_parser.go -o 'bazel-out/host/bin/external/com_github_golang_protobuf/proto/linux_amd64_race_stripped/go_default_library~/github.com/golang/protobuf/proto.a' -trimpath . -I . -- -p github.com/golang/protobuf/proto -race)
external/com_github_golang_protobuf/proto/clone.go:38:2: can't find import: "log"
2018/02/09 07:16:01 error running compiler: exit status 1
```

--- a/go/private/context.bzl
+++ b/go/private/context.bzl
@@ -198,10 +198,9 @@ def go_context(ctx, attr=None):
     builders = builders[GoBuilders]
   else:
     builders = GoBuilders(compile=None, link=None)
-  bootstrap = builders.compile == None
 
   context_data = attr._go_context_data
-  mode = get_mode(ctx, bootstrap, toolchain, context_data)
+  mode = get_mode(ctx, toolchain, context_data)
   root, binary = _get_go_binary(context_data)
 
   stdlib = getattr(attr, "_stdlib", None)

--- a/go/private/context.bzl
+++ b/go/private/context.bzl
@@ -198,9 +198,10 @@ def go_context(ctx, attr=None):
     builders = builders[GoBuilders]
   else:
     builders = GoBuilders(compile=None, link=None)
+  host_only = getattr(attr, "_hostonly", False)
 
   context_data = attr._go_context_data
-  mode = get_mode(ctx, toolchain, context_data)
+  mode = get_mode(ctx, host_only, toolchain, context_data)
   root, binary = _get_go_binary(context_data)
 
   stdlib = getattr(attr, "_stdlib", None)

--- a/go/private/mode.bzl
+++ b/go/private/mode.bzl
@@ -54,9 +54,10 @@ def _ternary(*values):
     fail("Invalid value {}".format(v))
   fail("_ternary failed to produce a final result from {}".format(values))
 
-def get_mode(ctx, go_toolchain, go_context_data):
+def get_mode(ctx, host_only, go_toolchain, go_context_data):
   # We always have to  use the pure stdlib in cross compilation mode
   force_pure = "on" if go_toolchain.cross_compile else "auto"
+  force_race = "off" if host_only else "auto"
 
   static = _ternary(
       getattr(ctx.attr, "static", None),
@@ -64,7 +65,7 @@ def get_mode(ctx, go_toolchain, go_context_data):
   )
   race = _ternary(
       getattr(ctx.attr, "race", None),
-      getattr(ctx.attr, "_race", None),
+      force_race,
       "race" in ctx.features,
   )
   msan = _ternary(

--- a/go/private/mode.bzl
+++ b/go/private/mode.bzl
@@ -54,10 +54,9 @@ def _ternary(*values):
     fail("Invalid value {}".format(v))
   fail("_ternary failed to produce a final result from {}".format(values))
 
-def get_mode(ctx, bootstrap, go_toolchain, go_context_data):
+def get_mode(ctx, go_toolchain, go_context_data):
   # We always have to  use the pure stdlib in cross compilation mode
   force_pure = "on" if go_toolchain.cross_compile else "auto"
-  force_race = "off" if bootstrap else "auto"
 
   static = _ternary(
       getattr(ctx.attr, "static", None),
@@ -65,7 +64,7 @@ def get_mode(ctx, bootstrap, go_toolchain, go_context_data):
   )
   race = _ternary(
       getattr(ctx.attr, "race", None),
-      force_race,
+      getattr(ctx.attr, "_race", None),
       "race" in ctx.features,
   )
   msan = _ternary(

--- a/go/private/rules/binary.bzl
+++ b/go/private/rules/binary.bzl
@@ -159,6 +159,7 @@ go_tool_binary = go_rule(
         "linkstamp": attr.string(),
         "x_defs": attr.string_dict(),
         "linkmode": attr.string(values=LINKMODES, default=LINKMODE_NORMAL),
+        "_race": attr.string(default="off"),
     },
     executable = True,
 )

--- a/go/private/rules/binary.bzl
+++ b/go/private/rules/binary.bzl
@@ -159,7 +159,7 @@ go_tool_binary = go_rule(
         "linkstamp": attr.string(),
         "x_defs": attr.string_dict(),
         "linkmode": attr.string(values=LINKMODES, default=LINKMODE_NORMAL),
-        "_race": attr.string(default="off"),
+        "_hostonly": attr.bool(default=True),
     },
     executable = True,
 )

--- a/go/private/rules/rule.bzl
+++ b/go/private/rules/rule.bzl
@@ -29,8 +29,6 @@ def go_rule(implementation, attrs={}, toolchains=[], bootstrap=False, **kwargs):
   if not bootstrap:
     attrs["_stdlib"] = attr.label(default = Label("@io_bazel_rules_go//:stdlib"), aspects = aspects)
     attrs["_builders"] = attr.label(default = Label("@io_bazel_rules_go//:builders"))
-  else:
-    attrs["_builders"] = attr.label(default = None)
 
   return rule(
       implementation = implementation,

--- a/tests/legacy/race/BUILD.bazel
+++ b/tests/legacy/race/BUILD.bazel
@@ -52,9 +52,10 @@ bazel_test(
     name = "race_feature",
     args = ["--features=race"],
     check = """
-if [ "$result" -eq 0 ]; then
-  echo "error: test succeeded unexpectedly" >&2
-  result=1
+if [ "$result" -ne 3 ]; then
+  echo "error: bazel test failure expected (code 3). Got code $result" >&2
+  cat bazel-output.txt
+  exit 1
 else
   result=0
 fi

